packages/block-editor/src/store/selectors.js

```diff
@@ -1791,6 +1792,32 @@ export const __experimentalGetAllowedBlocks = createSelector(
 	]
 );
 
+const checkAllowListRecursive = ( blocks, allowedBlockTypes ) => {
+	if ( isBoolean( allowedBlockTypes ) ) {
+		return allowedBlockTypes;
+	}
+
+	const blocksQueue = [ ...blocks ];
+	while ( blocksQueue.length > 0 ) {
+		const block = blocksQueue.shift();
+
+		const isAllowed = checkAllowList(
+			allowedBlockTypes,
+			block.name || block.blockName,
+			true
+		);
+		if ( ! isAllowed ) {
+			return false;
+		}
+
+		block.innerBlocks?.forEach( ( innerBlock ) => {
+			blocksQueue.push( innerBlock );
+		} );
+	}
+
+	return true;
+};
```

Comment:
We use the `checkAllowList` in `canInsertBlockTypeUnmemoized`. It was never a recursive function.

Reply:
sure, I get that we might use this already, but is there a reason they shouldn't be recursive? It sounds like you identified an issue _because_ it wasn't recursive, and do we end up with special rules for different circumstances?

Reply:
~It's not an issue. In `canInsertBlockTypeUnmemoized` only works with block types, so it can't work in a recursive way.

Patterns are special in this case, since they contain multiple block types. We don't have any other way to insert multiple kinds of block types in a single action.~

That was my thought, but reusable blocks are pretty similar 🤔 Like saving a Group as reusable block and then inserting it. This needs more testing and we shouldn't do such a huge change to the core in this PR.